src/main/java/frc/subsystem/CargoTracker.java

```diff
@@ -0,0 +1,201 @@
+    ArrayList<CargoPosition> redCargoPositions = new ArrayList<>(30);
+    private static final double[] NO_DETECTION = {0};
+    private static final Vector3d CAMERA_RELATIVE_POSITION = new Vector3d(0, 0, 0);
+    public static final double CARGO_RADIUS = 5;
+
+    private static final double CARGO_RADIUS_SQUARED = CARGO_RADIUS * CARGO_RADIUS;
+
+    private static final Vector2d[] CENTER_FIELD_GEOMETRY = {
+            new Vector2d(6.9270, -0.3846),
+            new Vector2d(7.2446, -0.2393),
+            new Vector2d(7.6512, 0.84890),
+            new Vector2d(7.5067, 1.18470),
+            new Vector2d(7.8602, 1.33040),
+            new Vector2d(8.0020, 1.00600),
+            new Vector2d(9.1144, 0.59000),
+            new Vector2d(9.4292, 0.72970),
+            new Vector2d(9.5957, 0.39600),
+            new Vector2d(9.2665, 0.25170),
+            new Vector2d(8.8336, -0.8476),
+            new Vector2d(8.9712, -1.1572),
+            new Vector2d(8.6237, -1.3166),
+            new Vector2d(8.4980, -1.0302),
+            new Vector2d(7.3787, -0.5838)
+    };
+
+
+    private void updateCargoPositions() {
+        double timestamp = Timer.getFPGATimestamp() - (latencyTable.getDouble(0) / 1000) - Constants.REALSENSE_NETWORK_LATENCY;
+
+        Optional<Pose2d> optionalRobotPose = RobotTracker.getInstance().getPoseAtTime(timestamp);
+        if (optionalRobotPose.isEmpty()) {
+            return;
+        }
+
+        final @NotNull Vector3d[] detectedBlueCargoPositions = getCargoPositions(blueCargoTable.getDoubleArray(NO_DETECTION));
+        final @NotNull Vector3d[] detectedRedCargoPositions = getCargoPositions(redCargoTable.getDoubleArray(NO_DETECTION));
+        Pose2d robotPose2d = optionalRobotPose.get();
+        Vector3d robotPose = new Vector3d(robotPose2d.getTranslation().getX(), 0, robotPose2d.getTranslation().getY());
+
+        AxisAngle4d robotRotation = new AxisAngle4d(robotPose2d.getRotation().getRadians(), 0, 1, 0);
+        Vector3d cameraPosition = new Vector3d(CAMERA_RELATIVE_POSITION);
+        robotRotation.transform(cameraPosition);
+        cameraPosition.add(robotPose);
+
+        // Make the cargo positions relative to the field
+
+        updateCargoPositions(timestamp, detectedBlueCargoPositions, robotRotation, cameraPosition, blueCargoPositions);
+        updateCargoPositions(timestamp, detectedRedCargoPositions, robotRotation, cameraPosition, redCargoPositions);
+        lastTimestamp = timestamp;
+    }
+
+    private void updateCargoPositions(double timestamp, @NotNull Vector3d @NotNull [] detectedCargoPositions,
+                                      @NotNull AxisAngle4d robotRotation, @NotNull Vector3d cameraPosition,
+                                      @NotNull ArrayList<CargoPosition> cargoPositions) {
+        for (Vector3d detectedCargoPosition : detectedCargoPositions) {
+            INTAKE_CAMERA_ROTATION.transform(robotRotation.transform(detectedCargoPosition)).add(cameraPosition);
```

Comment:
is detected cargo position in the intake camera reference frame? if that is true shouldnt you be applying these in reverse order to get the field relative position:
cargo_wrt_field = robotRotation ^-1.transform(intake_camera_rotation^-1.transform(detectedCargoPosition) - cameraPosition) 

maybe i am misunderstanding what is going on here

Reply:
Does the order we apply the rotations really matter? They're on their own axis's. Also, we shouldn't we add the cameraPosition to make the position relative to the field? (I changed the name of cameraPosition to cameraPositionRelativeToField to make it clearer what it is.)

I don't think we want to find the inverse of the robot rotation, right?